File: SystemMonitor/MonitoringService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Management;
using System.Net.Http;
using System.Threading.Tasks;
using System.Windows.Threading;
using Microsoft.Win32;
using Newtonsoft.Json;
using System.Linq;

namespace SystemMonitor
{
    public class MonitoringService
    {
        private readonly HttpClient _httpClient;
        private readonly AppConfig _config;
        private ManagementEventWatcher _usbWatcher;
        private ManagementEventWatcher _processWatcher;
        private List<FileSystemWatcher> _fileWatchers;
        private bool _isMonitoring;
        private UsbBlockingService _usbBlockingService;
        private GoogleSheetsManager _sheetsManager;
        private UninstallDetectionService _uninstallDetectionService;

        public event EventHandler<ActivityEventArgs> ActivityDetected;

        public MonitoringService(AppConfig config = null)
        {
            _httpClient = new HttpClient();
            _config = config ?? AppConfig.LoadFromFile();
            _fileWatchers = new List<FileSystemWatcher>();
            _isMonitoring = false;
            
            // Initialize USB blocking if enabled
            if (_config.UsbBlockingSettings.EnableUsbBlocking && 
                !string.IsNullOrEmpty(_config.UsbBlockingSettings.GoogleSheetsApiKey) &&
                !string.IsNullOrEmpty(_config.UsbBlockingSettings.GoogleSheetsSpreadsheetId))
            {
                _sheetsManager = new GoogleSheetsManager(
                    _config.UsbBlockingSettings.GoogleSheetsApiKey,
                    _config.UsbBlockingSettings.GoogleSheetsSpreadsheetId,
                    _config.UsbBlockingSettings.GoogleSheetsRange);
                
                _usbBlockingService = new UsbBlockingService(_sheetsManager, true);
                _usbBlockingService.UsbBlocked += OnUsbBlocked;
            }

            // Initialize uninstall detection
            if (!string.IsNullOrEmpty(_config.N8nWebhookUrl))
            {
                _uninstallDetectionService = new UninstallDetectionService(_config.N8nWebhookUrl);
                _uninstallDetectionService.InitializeUninstallDetection();
            }
        }

        public void StartMonitoring()
        {
            if (_isMonitoring) return;

            _isMonitoring = true;
            
            // Start USB blocking first if enabled
            if (_usbBlockingService != null)
            {
                _usbBlockingService.StartBlocking();
            }
            
            if (_config.MonitoringSettings.EnableUsbMonitoring)
                StartUsbMonitoring();
                
            if (_config.MonitoringSettings.EnableAppInstallationMonitoring || 
                _config.MonitoringSettings.EnableBlacklistedAppMonitoring)
                StartProcessMonitoring();
                
            if (_config.MonitoringSettings.EnableFileTransferMonitoring)
                StartFileTransferMonitoring();
                
            if (_config.MonitoringSettings.EnableNetworkMonitoring)
                StartNetworkMonitoring();
        }

        public void StopMonitoring()
        {
            if (!_isMonitoring) return;

            _isMonitoring = false;
            _usbWatcher?.Stop();
            _processWatcher?.Stop();
            _usbBlockingService?.StopBlocking();
            
            foreach (var watcher in _fileWatchers)
            {
                watcher?.Dispose();
            }
            _fileWatchers.Clear();
        }

        public async Task SendUninstallNotification()
        {
            if (_uninstallDetectionService != null)
            {
                await _uninstallDetectionService.SendUninstallNotification();
            }
        }

        private void OnUsbBlocked(object sender, UsbBlockingEventArgs e)
        {
            var activity = new ActivityEventArgs
            {
                Type = ActivityType.UsbBlocked,
                Description = $"USB device blocked: {e.DeviceId} - {e.Reason}",
                Timestamp = e.Timestamp,
                Severity = ActivitySeverity.High,
                Details = new Dictionary<string, string>
                {
                    ["DeviceID"] = e.DeviceId,
                    ["Reason"] = e.Reason,
                    ["Blocked"] = e.Blocked.ToString()
                }
            };

            OnActivityDetected(activity);
        }

        private void StartUsbMonitoring()
        {
            try
            {
                var query = new WqlEventQuery("SELECT * FROM Win32_DeviceChangeEvent WHERE EventType = 2 or EventType = 3");
                _usbWatcher = new ManagementEventWatcher(query);
                _usbWatcher.EventArrived += (sender, e) =>
                {
                    var eventType = e.NewEvent["EventType"];
                    var deviceId = e.NewEvent["DeviceID"];
                    
                    var activity = new ActivityEventArgs
                    {
                        Type = ActivityType.UsbDrive,
                        Description = $"USB device {(eventType.ToString() == "2" ? "connected" : "disconnected")}: {deviceId}",
                        Timestamp = DateTime.Now,
                        Severity = ActivitySeverity.Medium,
                        Details = new Dictionary<string, string>
                        {
                            ["EventType"] = eventType?.ToString() ?? "Unknown",
                            ["DeviceID"] = deviceId?.ToString() ?? "Unknown"
                        }
                    };

                    OnActivityDetected(activity);
                };
                _usbWatcher.Start();
            }
            catch (Exception ex)
            {
                OnActivityDetected(new ActivityEventArgs
                {
                    Type = ActivityType.System,
                    Description = $"USB monitoring error: {ex.Message}",
                    Timestamp = DateTime.Now,
                    Severity = ActivitySeverity.High
                });
            }
        }

        private void StartProcessMonitoring()
        {
            try
            {
                var query = new WqlEventQuery("SELECT * FROM Win32_ProcessStartTrace");
                _processWatcher = new ManagementEventWatcher(query);
                _processWatcher.EventArrived += (sender, e) =>
                {
                    var processName = e.NewEvent["ProcessName"]?.ToString();
                    var processId = e.NewEvent["ProcessID"]?.ToString();
                    var commandLine = e.NewEvent["CommandLine"]?.ToString();

                    if (!string.IsNullOrEmpty(processName))
                    {
                        // Check for blacklisted applications
                        if (_config.MonitoringSettings.EnableBlacklistedAppMonitoring && 
                            IsBlacklistedApp(processName))
                        {
                            var activity = new ActivityEventArgs
                            {
                                Type = ActivityType.BlacklistedApp,
                                Description = $"Blacklisted application detected: {processName} (PID: {processId})",
                                Timestamp = DateTime.Now,
                                Severity = ActivitySeverity.High,
                                Details = new Dictionary<string, string>
                                {
                                    ["ProcessName"] = processName,
                                    ["ProcessID"] = processId,
                                    ["CommandLine"] = commandLine
                                }
                            };

                            OnActivityDetected(activity);
                        }

                        // Check for installation activities
                        if (_config.MonitoringSettings.EnableAppInstallationMonitoring && 
                            IsInstallationProcess(processName, commandLine))
                        {
                            var activity = new ActivityEventArgs
                            {
                                Type = ActivityType.AppInstallation,
                                Description = $"Application installation detected: {processName}",
                                Timestamp = DateTime.Now,
                                Severity = ActivitySeverity.Medium,
                                Details = new Dictionary<string, string>
                                {
                                    ["ProcessName"] = processName,
                                    ["ProcessID"] = processId,
                                    ["CommandLine"] = commandLine
                                }
                            };

                            OnActivityDetected(activity);
                        }
                    }
                };
                _processWatcher.Start();
            }
            catch (Exception ex)
            {
                OnActivityDetected(new ActivityEventArgs
                {
                    Type = ActivityType.System,
                    Description = $"Process monitoring error: {ex.Message}",
                    Timestamp = DateTime.Now,
                    Severity = ActivitySeverity.High
                });
            }
        }

        private void StartFileTransferMonitoring()
        {
            try
            {
                // Monitor common USB drive letters
                var driveLetters = new[] { "D:", "E:", "F:", "G:", "H:", "I:", "J:", "K:", "L:", "M:", "N:", "O:", "P:" };
                
                foreach (var drive in driveLetters)
                {
                    if (Directory.Exists(drive))
                    {
                        var watcher = new FileSystemWatcher(drive)
                        {
                            IncludeSubdirectories = true,
                            EnableRaisingEvents = true,
                            NotifyFilter = NotifyFilters.FileName | NotifyFilters.DirectoryName | NotifyFilters.Size
                        };

                        watcher.Created += (sender, e) =>
                        {
                            var activity = new ActivityEventArgs
                            {
                                Type = ActivityType.FileTransfer,
                                Description = $"File created on USB drive: {e.FullPath}",
                                Timestamp = DateTime.Now,
                                Severity = ActivitySeverity.Medium,
                                Details = new Dictionary<string, string>
                                {
                                    ["FilePath"] = e.FullPath,
                                    ["Drive"] = drive,
                                    ["FileSize"] = GetFileSize(e.FullPath),
                                    ["EventType"] = "Created"
                                }
                            };

                            OnActivityDetected(activity);
                        };

                        watcher.Changed += (sender, e) =>
                        {
                            var activity = new ActivityEventArgs
                            {
                                Type = ActivityType.FileTransfer,
                                Description = $"File modified on USB drive: {e.FullPath}",
                                Timestamp = DateTime.Now,
                                Severity = ActivitySeverity.Low,
                                Details = new Dictionary<string, string>
                                {
                                    ["FilePath"] = e.FullPath,
                                    ["Drive"] = drive,
                                    ["FileSize"] = GetFileSize(e.FullPath),
                                    ["EventType"] = "Modified"
                                }
                            };

                            OnActivityDetected(activity);
                        };

                        watcher.Deleted += (sender, e) =>
                        {
                            var activity = new ActivityEventArgs
                            {
                                Type = ActivityType.FileTransfer,
                                Description = $"File deleted from USB drive: {e.FullPath}",
                                Timestamp = DateTime.Now,
                                Severity = ActivitySeverity.Low,
                                Details = new Dictionary<string, string>
                                {
                                    ["FilePath"] = e.FullPath,
                                    ["Drive"] = drive,
                                    ["EventType"] = "Deleted"
                                }
                            };

                            OnActivityDetected(activity);
                        };

                        _fileWatchers.Add(watcher);
                    }
                }
            }
            catch (Exception ex)
            {
                OnActivityDetected(new ActivityEventArgs
                {
                    Type = ActivityType.System,
                    Description = $"File transfer monitoring error: {ex.Message}",
                    Timestamp = DateTime.Now,
                    Severity = ActivitySeverity.High
                });
            }
        }

        private void StartNetworkMonitoring()
        {
            // Monitor network connections using WMI
            try
            {
                var query = new WqlEventQuery("SELECT * FROM Win32_NetworkAdapterConfiguration WHERE IPEnabled = true");
                var networkWatcher = new ManagementEventWatcher(query);
                networkWatcher.EventArrived += (sender, e) =>
                {
                    // Check for suspicious network activity
                    var activity = new ActivityEventArgs
                    {
                        Type = ActivityType.NetworkActivity,
                        Description = "Network configuration change detected",
                        Timestamp = DateTime.Now,
                        Severity = ActivitySeverity.Medium,
                        Details = new Dictionary<string, string>
                        {
                            ["EventType"] = "NetworkConfigurationChange"
                        }
                    };

                    OnActivityDetected(activity);
                };
                networkWatcher.Start();
            }
            catch (Exception ex)
            {
                OnActivityDetected(new ActivityEventArgs
                {
                    Type = ActivityType.System,
                    Description = $"Network monitoring error: {ex.Message}",
                    Timestamp = DateTime.Now,
                    Severity = ActivitySeverity.High
                });
            }
        }

        private bool IsBlacklistedApp(string processName)
        {
            return _config.BlacklistedApps.Contains(processName.ToLower());
        }

        private bool IsInstallationProcess(string processName, string commandLine)
        {
            var installKeywords = new[] { "install", "setup", "msi", "exe", "installer", "uninstall" };
            var processNameLower = processName.ToLower();
            var commandLineLower = commandLine?.ToLower() ?? "";

            return installKeywords.Any(keyword => 
                processNameLower.Contains(keyword) || commandLineLower.Contains(keyword));
        }

        private string GetFileSize(string filePath)
        {
            try
            {
                if (File.Exists(filePath))
                {
                    var fileInfo = new FileInfo(filePath);
                    return fileInfo.Length.ToString();
                }
                return "Unknown";
            }
            catch
            {
                return "Unknown";
            }
        }

        public async Task SendToN8n(ActivityEventArgs activity)
        {
            if (string.IsNullOrEmpty(_config.N8nWebhookUrl) || !_config.MonitoringSettings.SendToN8n) 
                return;

            for (int attempt = 1; attempt <= _config.MonitoringSettings.N8nRetryAttempts; attempt++)
            {
                try
                {
                    var payload = new
                    {
                        timestamp = activity.Timestamp.ToString("yyyy-MM-dd HH:mm:ss"),
                        type = activity.Type.ToString(),
                        description = activity.Description,
                        severity = activity.Severity.ToString(),
                        details = activity.Details,
                        computer = Environment.MachineName,
                        user = Environment.UserName,
                        config = new
                        {
                            logLevel = _config.MonitoringSettings.LogLevel,
                            maxLogEntries = _config.MonitoringSettings.MaxLogEntries
                        }
                    };

                    var json = JsonConvert.SerializeObject(payload);
                    var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
                    
                    var response = await _httpClient.PostAsync(_config.N8nWebhookUrl, content);
                    
                    if (!response.IsSuccessStatusCode)
                    {
                        throw new Exception($"HTTP {response.StatusCode}: {response.ReasonPhrase}");
                    }

                    // Success, break out of retry loop
                    break;
                }
                catch (Exception ex)
                {
                    if (attempt == _config.MonitoringSettings.N8nRetryAttempts)
                    {
                        OnActivityDetected(new ActivityEventArgs
                        {
                            Type = ActivityType.System,
                            Description = $"Failed to send to N8N after {attempt} attempts: {ex.Message}",
                            Timestamp = DateTime.Now,
                            Severity = ActivitySeverity.High
                        });
                    }
                    else
                    {
                        // Wait before retry
                        await Task.Delay(_config.MonitoringSettings.N8nRetryDelayMs);
                    }
                }
            }
        }

        protected virtual void OnActivityDetected(ActivityEventArgs e)
        {
            ActivityDetected?.Invoke(this, e);
        }

        public void Dispose()
        {
            StopMonitoring();
            _httpClient?.Dispose();
            _usbWatcher?.Dispose();
            _processWatcher?.Dispose();
            _usbBlockingService?.Dispose();
            _sheetsManager?.Dispose();
            _uninstallDetectionService?.Dispose();
            
            foreach (var watcher in _fileWatchers)
            {
                watcher?.Dispose();
            }
        }
    }

    public class ActivityEventArgs : EventArgs
    {
        public ActivityType Type { get; set; }
        public string Description { get; set; }
        public DateTime Timestamp { get; set; }
        public ActivitySeverity Severity { get; set; }
        public Dictionary<string, string> Details { get; set; } = new Dictionary<string, string>();
    }

    public enum ActivityType
    {
        UsbDrive,
        FileTransfer,
        AppInstallation,
        BlacklistedApp,
        NetworkActivity,
        System,
        UsbBlocked
    }

    public enum ActivitySeverity
    {
        Low,
        Medium,
        High,
        Critical
    }
} 

File: MacSystemMonitor/MonitoringService.swift
import Foundation
import IOKit
import SystemConfiguration

class MonitoringService: ObservableObject {
    @Published var monitoringStatus: MonitoringStatus = .stopped
    @Published var usbBlockingStatus: UsbBlockingStatus = .disabled
    @Published var activities: [ActivityEvent] = []
    
    private let config: AppConfig
    private var usbBlockingService: UsbBlockingService?
    private var sheetsManager: GoogleSheetsManager?
    private var uninstallDetectionService: UninstallDetectionService?
    private var isMonitoring: Bool = false
    
    weak var delegate: MonitoringServiceDelegate?
    
    init(config: AppConfig = AppConfig.loadFromFile()) {
        self.config = config
        
        // Initialize USB blocking if enabled
        if config.usbBlockingSettings.enableUsbBlocking &&
           !config.usbBlockingSettings.googleSheetsApiKey.isEmpty &&
           !config.usbBlockingSettings.googleSheetsSpreadsheetId.isEmpty {
            
            sheetsManager = GoogleSheetsManager(
                apiKey: config.usbBlockingSettings.googleSheetsApiKey,
                spreadsheetId: config.usbBlockingSettings.googleSheetsSpreadsheetId,
                range: config.usbBlockingSettings.googleSheetsRange,
                cacheExpirationMinutes: config.usbBlockingSettings.cacheExpirationMinutes
            )
            
            usbBlockingService = UsbBlockingService(sheetsManager: sheetsManager!, enableBlocking: true)
            usbBlockingService?.delegate = self
        }
        
        // Initialize uninstall detection
        if !config.n8nWebhookUrl.isEmpty {
            uninstallDetectionService = UninstallDetectionService(n8nWebhookUrl: config.n8nWebhookUrl)
            uninstallDetectionService?.initializeUninstallDetection()
            uninstallDetectionService?.setupProcessExitHandling()
        }
    }
    
    // MARK: - Monitoring Control
    
    func startMonitoring() {
        guard !isMonitoring else { return }
        
        isMonitoring = true
        monitoringStatus = .running
        
        // Start USB blocking first if enabled
        if let usbService = usbBlockingService {
            usbService.startBlocking()
            usbBlockingStatus = .enabled
        }
        
        // Start various monitoring components
        if config.monitoringSettings.enableUsbMonitoring {
            startUsbMonitoring()
        }
        
        if config.monitoringSettings.enableAppInstallationMonitoring ||
           config.monitoringSettings.enableBlacklistedAppMonitoring {
            startProcessMonitoring()
        }
        
        if config.monitoringSettings.enableFileTransferMonitoring {
            startFileTransferMonitoring()
        }
        
        if config.monitoringSettings.enableNetworkMonitoring {
            startNetworkMonitoring()
        }
        
        print("Monitoring started")
    }
    
    func stopMonitoring() {
        guard isMonitoring else { return }
        
        isMonitoring = false
        monitoringStatus = .stopped
        
        usbBlockingService?.stopBlocking()
        usbBlockingStatus = .disabled
        
        print("Monitoring stopped")
    }
    
    // MARK: - USB Monitoring
    
    private func startUsbMonitoring() {
        // USB monitoring is handled by UsbBlockingService
        print("USB monitoring started")
    }
    
    // MARK: - Process Monitoring
    
    private func startProcessMonitoring() {
        // Monitor for new processes using system_profiler and ps
        let queue = DispatchQueue.global(qos: .background)
        queue.async {
            self.monitorProcesses()
        }
    }
    
    private func monitorProcesses() {
        let task = Process()
        task.launchPath = "/bin/ps"
        task.arguments = ["-ax", "-o", "pid,comm,args"]
        
        let pipe = Pipe()
        task.standardOutput = pipe
        
        do {
            try task.run()
            task.waitUntilExit()
            
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            if let output = String(data: data, encoding: .utf8) {
                let lines = output.components(separatedBy: .newlines)
                
                for line in lines {
                    let components = line.components(separatedBy: .whitespaces)
                    if components.count >= 2 {
                        let processName = components[1]
                        let commandLine = line
                        
                        // Check for blacklisted applications
                        if config.monitoringSettings.enableBlacklistedAppMonitoring &&
                           isBlacklistedApp(processName) {
                            
                            let activity = ActivityEvent(
                                type: .blacklistedApp,
                                description: "Blacklisted application detected: \(processName)",
                                severity: .high,
                                details: [
                                    "ProcessName": processName,
                                    "CommandLine": commandLine
                                ]
                            )
                            
                            DispatchQueue.main.async {
                                self.addActivity(activity)
                            }
                        }
                        
                        // Check for installation activities
                        if config.monitoringSettings.enableAppInstallationMonitoring &&
                           isInstallationProcess(processName, commandLine: commandLine) {
                            
                            let activity = ActivityEvent(
                                type: .appInstallation,
                                description: "Application installation detected: \(processName)",
                                severity: .medium,
                                details: [
                                    "ProcessName": processName,
                                    "CommandLine": commandLine
                                ]
                            )
                            
                            DispatchQueue.main.async {
                                self.addActivity(activity)
                            }
                        }
                    }
                }
            }
        } catch {
            print("Failed to monitor processes: \(error)")
        }
        
        // Schedule next check
        DispatchQueue.global().asyncAfter(deadline: .now() + 10.0) {
            self.monitorProcesses()
        }
    }
    
    private func isBlacklistedApp(_ processName: String) -> Bool {
        return config.blacklistedApps.contains { processName.lowercased().contains($0.lowercased()) }
    }
    
    private func isInstallationProcess(_ processName: String, commandLine: String) -> Bool {
        let installKeywords = ["install", "setup", "pkg", "dmg", "installer", "uninstall"]
        let processNameLower = processName.lowercased()
        let commandLineLower = commandLine.lowercased()
        
        return installKeywords.contains { keyword in
            processNameLower.contains(keyword) || commandLineLower.contains(keyword)
        }
    }
    
    // MARK: - File Transfer Monitoring
    
    private func startFileTransferMonitoring() {
        // Monitor common external storage locations
        let externalPaths = [
            "/Volumes",
            NSHomeDirectory() + "/Desktop",
            NSHomeDirectory() + "/Downloads"
        ]
        
        for path in externalPaths {
            monitorDirectory(path)
        }
    }
    
    private func monitorDirectory(_ path: String) {
        let queue = DispatchQueue.global(qos: .background)
        queue.async {
            self.watchDirectory(path)
        }
    }
    
    private func watchDirectory(_ path: String) {
        let fileManager = FileManager.default
        
        guard fileManager.fileExists(atPath: path) else { return }
        
        // Get initial file list
        var initialFiles: Set<String> = []
        do {
            let files = try fileManager.contentsOfDirectory(atPath: path)
            initialFiles = Set(files)
        } catch {
            print("Failed to get initial file list for \(path): \(error)")
            return
        }
        
        // Monitor for changes
        DispatchQueue.global().asyncAfter(deadline: .now() + 5.0) {
            do {
                let currentFiles = try fileManager.contentsOfDirectory(atPath: path)
                let currentFileSet = Set(currentFiles)
                
                // Check for new files
                let newFiles = currentFileSet.subtracting(initialFiles)
                for file in newFiles {
                    let filePath = path + "/" + file
                    let activity = ActivityEvent(
                        type: .fileTransfer,
                        description: "File created: \(file)",
                        severity: .medium,
                        details: [
                            "FilePath": filePath,
                            "Directory": path,
                            "EventType": "Created"
                        ]
                    )
                    
                    DispatchQueue.main.async {
                        self.addActivity(activity)
                    }
                }
                
                // Check for deleted files
                let deletedFiles = initialFiles.subtracting(currentFileSet)
                for file in deletedFiles {
                    let activity = ActivityEvent(
                        type: .fileTransfer,
                        description: "File deleted: \(file)",
                        severity: .low,
                        details: [
                            "FilePath": path + "/" + file,
                            "Directory": path,
                            "EventType": "Deleted"
                        ]
                    )
                    
                    DispatchQueue.main.async {
                        self.addActivity(activity)
                    }
                }
                
                // Update initial files for next check
                initialFiles = currentFileSet
            } catch {
                print("Failed to monitor directory \(path): \(error)")
            }
            
            // Continue monitoring
            self.watchDirectory(path)
        }
    }
    
    // MARK: - Network Activity Monitoring
    
    private func startNetworkMonitoring() {
        // Monitor network connections using netstat
        let queue = DispatchQueue.global(qos: .background)
        queue.async {
            self.monitorNetworkConnections()
        }
    }
    
    private func monitorNetworkConnections() {
        let task = Process()
        task.launchPath = "/usr/bin/netstat"
        task.arguments = ["-an"]
        
        let pipe = Pipe()
        task.standardOutput = pipe
        
        do {
            try task.run()
            task.waitUntilExit()
            
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            if let output = String(data: data, encoding: .utf8) {
                let lines = output.components(separatedBy: .newlines)
                
                for line in lines {
                    if line.contains("ESTABLISHED") {
                        // Check for suspicious connections
                        for domain in config.suspiciousDomains {
                            if line.contains(domain) {
                                let activity = ActivityEvent(
                                    type: .networkActivity,
                                    description: "Suspicious network connection detected",
                                    severity: .high,
                                    details: [
                                        "Connection": line,
                                        "SuspiciousDomain": domain
                                    ]
                                )
                                
                                DispatchQueue.main.async {
                                    self.addActivity(activity)
                                }
                                break
                            }
                        }
                    }
                }
            }
        } catch {
            print("Failed to monitor network connections: \(error)")
        }
        
        // Schedule next check
        DispatchQueue.global().asyncAfter(deadline: .now() + 30.0) {
            self.monitorNetworkConnections()
        }
    }
    
    // MARK: - Activity Management
    
    private func addActivity(_ activity: ActivityEvent) {
        DispatchQueue.main.async {
            self.activities.insert(activity, at: 0)
            
            // Limit the number of activities
            if self.activities.count > self.config.monitoringSettings.maxLogEntries {
                self.activities = Array(self.activities.prefix(self.config.monitoringSettings.maxLogEntries))
            }
            
            // Send to N8N if enabled
            if self.config.monitoringSettings.sendToN8n {
                Task {
                    await self.sendToN8n(activity)
                }
            }
            
            // Notify delegate
            self.delegate?.activityDetected(activity)
        }
    }
    
    // MARK: - N8N Integration
    
    private func sendToN8n(_ activity: ActivityEvent) async {
        guard !config.n8nWebhookUrl.isEmpty else { return }
        
        for attempt in 1...config.monitoringSettings.n8nRetryAttempts {
            do {
                let payload = N8nPayload(from: activity)
                let jsonData = try JSONEncoder().encode(payload)
                
                var request = URLRequest(url: URL(string: config.n8nWebhookUrl)!)
                request.httpMethod = "POST"
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.httpBody = jsonData
                
                let (data, response) = try await URLSession.shared.data(for: request)
                
                if let httpResponse = response as? HTTPURLResponse {
                    if httpResponse.statusCode == 200 {
                        print("Activity sent to N8N successfully")
                        break
                    } else {
                        throw NSError(domain: "N8N", code: httpResponse.statusCode, userInfo: nil)
                    }
                }
            } catch {
                print("Failed to send to N8N (attempt \(attempt)): \(error)")
                
                if attempt < config.monitoringSettings.n8nRetryAttempts {
                    try? await Task.sleep(nanoseconds: UInt64(config.monitoringSettings.n8nRetryDelayMs * 1_000_000))
                }
            }
        }
    }
    
    // MARK: - Uninstall Notification
    
    func sendUninstallNotification() async {
        await uninstallDetectionService?.sendUninstallNotification()
    }
}

// MARK: - Monitoring Service Delegate

protocol MonitoringServiceDelegate: AnyObject {
    func activityDetected(_ activity: ActivityEvent)
}

// MARK: - USB Blocking Service Delegate Implementation

extension MonitoringService: UsbBlockingServiceDelegate {
    func usbDeviceBlocked(_ event: UsbBlockingEvent) {
        let activity = ActivityEvent(
            type: .usbBlocked,
            description: "USB device blocked: \(event.deviceName ?? event.deviceId)",
            severity: .high,
            details: [
                "DeviceID": event.deviceId,
                "Reason": event.reason,
                "Blocked": String(event.blocked)
            ]
        )
        
        addActivity(activity)
    }
    
    func usbDeviceRemoved(_ deviceInfo: UsbDeviceInfo) {
        let activity = ActivityEvent(
            type: .usbDrive,
            description: "USB device removed: \(deviceInfo.displayName)",
            severity: .medium,
            details: [
                "DeviceID": deviceInfo.deviceId,
                "DeviceName": deviceInfo.displayName
            ]
        )
        
        addActivity(activity)
    }
} 